types/decode.go

```diff
@@ -87,12 +87,25 @@ func Unmarshal(bs []byte) (result Type, err error) {
 		case typeFunction:
 			var decl rawdecl
 			if err = util.UnmarshalJSON(bs, &decl); err == nil {
-				var args []Type
-				if args, err = unmarshalSlice(decl.Args); err == nil {
-					var ret Type
-					if ret, err = Unmarshal(decl.Result); err == nil {
-						result = NewFunction(args, ret)
+				args, err := unmarshalSlice(decl.Args)
+				if err != nil {
+					return nil, err
+				}
+				var ret Type
+				if len(decl.Result) > 0 {
+					ret, err = Unmarshal(decl.Result)
+					if err != nil {
+						return nil, err
+					}
+				}
+				if len(decl.Variadic) == 0 {
+					result = NewFunction(args, ret)
+				} else {
+					varargs, err := Unmarshal(decl.Variadic)
+					if err != nil {
+						return nil, err
 					}
+					result = NewVariadicFunction(args, varargs, ret)
```

Comment:
[style nit] Semantically, this reads like _"If some condition is not the case, do the default thing. Otherwise, deal with it."_
I think I'd flip those branches...